pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java

```diff
@@ -79,128 +69,145 @@ public RoundRobinScheduler(long releaseTimeout) {
-    // we just store it in a set of seen mail and only check for it when hasNext
-    // is called.
-    //
-    // note that scenario (2) may cause a false-positive schedule where an operator
-    // chain gets scheduled for mail that it had already processed, in which case
-    // the operator chain will simply do nothing and get put back onto the queue.
-    // scenario (2) may additionally cause a memory leak - if onDataAvailable is
-    // called with an EOS block _while_ the operator chain is executing, the chain
-    // will consume the EOS block and computeReady() will never remove the mailbox
-    // from the _seenMail set.
-    //
-    // TODO: fix the memory leak by adding a close(opChain) callback
-    _seenMail.add(mailbox);
-    trace("got mail for " + mailbox);
-  }
-
-  @Override
-  public boolean hasNext() {
-    if (!_ready.isEmpty()) {
-      return true;
+  public void deregister(OpChain operatorChain) {
+    _lock.lock();
+    try {
+      _aliveChains.remove(operatorChain.getId());
+      // deregister can only be called if the OpChain is running, so remove from _runningChains.
+      _runningChains.remove(operatorChain.getId());
+      // it could be that the onDataAvailable callback was called when the OpChain was executing, in which case there
+      // could be a dangling entry in _seenMail.
+      _seenMail.remove(operatorChain.getId());
+    } finally {
+      _lock.unlock();
     }
-    computeReady();
-    return !_ready.isEmpty();
-  }
-
-  @Override
-  public OpChain next() {
-    OpChain op = _ready.poll();
-    trace("Polled " + op);
-    return op;
   }
 
   @Override
-  public int size() {
-    return _ready.size() + _available.size();
+  public void yield(OpChain operatorChain) {
+    long releaseTs = _releaseTimeout < 0 ? Long.MAX_VALUE : _ticker.get() + _releaseTimeout;
+    _lock.lock();
+    try {
+      _runningChains.remove(operatorChain.getId());
+      // It could be that this OpChain received data before it could be yielded completely. In that case, mark it ready
+      // to get it scheduled asap.
+      if (_seenMail.contains(operatorChain.getId())) {
+        _seenMail.remove(operatorChain.getId());
+        _ready.add(operatorChain);
+        return;
+      }
+      _available.put(operatorChain.getId(), releaseTs);
+    } finally {
+      _lock.unlock();
+    }
   }
 
   @Override
-  public void shutDown() {
-    if (_isShutDown) {
+  public void onDataAvailable(MailboxIdentifier mailbox) {
+    // TODO: Should we add an API in MailboxIdentifier to get the requestId?
+    OpChainId opChainId = new OpChainId(Long.parseLong(mailbox.getJobId().split("_")[0]),
+        mailbox.getReceiverStageId());
+    // If this chain isn't alive as per the scheduler, don't do anything. If the OpChain is registered after this, it
+    // will anyways be scheduled to run since new OpChains are run immediately.
+    if (!_aliveChains.containsKey(opChainId)) {
```

Comment:
isn't it possible that the _aliveChains is not registered (e.g. downstream stage gets execute faster and start ending mails) but the _aliveChain is not yet registered with the current stage)? what would happen then, say there's no more onDataAvailable trigger?

Reply:
Yeah so in that case once the OpChain finally gets registered in the receiver, since we directly add it to the ready queue, it will start execution asap.

The way I think about `onDataAvailable` callback in context of this scheduler implementation is that it is supposed to only signal that a yielded OpChain can be resumed. If the OpChain itself is not registered yet, we can ignore this. Once it gets scheduled it will run anyways.

Reply:
got it that make sense. so the onDataAvailable is kind of a quicker yield->ready than the periodic scheduler task